### PR TITLE
[GOBBLIN-1185]Enable datasetCleaner to emit kafka event and add config field in Ver…

### PR DESCRIPTION
**Pull request by ZihanLi58**
…sionFinderAndPolicy

Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [ ] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-1185] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1185

### Description
- [ ] Here are some details about my PR, including screenshots (if applicable):
Add dynamic config to enable DatasetCleaner to emit kafka event
Add config field in VersionFinderAndPolicy so that CleanableDataset are aware of the config for different partition.

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-modules/gobblin-kafka-09/src/test/java/org/apache/gobblin/kafka/writer/Kafka09DataWriterTest.java [hunk 1]
```diff
@@ -84,7 +84,7 @@ public void testStringSerialization()
     _kafkaTestHelper.provisionTopic(topic);
     Properties props = new Properties();
     props.setProperty(KafkaWriterConfigurationKeys.KAFKA_TOPIC, topic);
-    props.setProperty(KafkaWriterConfigurationKeys.KAFKA_PRODUCER_CONFIG_PREFIX+"bootstrap.servers", "localhost:" + _kafkaTestHelper.getKafkaServerPort());
+    props.setProperty(KafkaWriterConfigurationKeys.KAFKA_PRODUCER_CONFIG_PREFIX+"bootstrap.servers", "127.0.0.1:" + _kafkaTestHelper.getKafkaServerPort());
```
Comment (autumnust) [inline gobblin-modules/gobblin-kafka-09/src/test/java/org/apache/gobblin/kafka/writer/Kafka09DataWriterTest.java:87]: why this change?

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/retention/DatasetCleaner.java [hunk 2]
```diff
@@ -88,33 +94,43 @@
...
-    State state = new State(props);
+    Properties properties = new Properties();
+    properties.putAll(props);
+    // load dynamic configuration and add them to the job properties
+    Config propsAsConfig = ConfigUtils.propertiesToConfig(props);
+    DynamicConfigGenerator dynamicConfigGenerator =
+        DynamicConfigGeneratorFactory.createDynamicConfigGenerator(propsAsConfig);
+    Config dynamicConfig = dynamicConfigGenerator.generateDynamicConfig(propsAsConfig);
+    for (Map.Entry<String, ConfigValue> entry : dynamicConfig.entrySet()) {
+      properties.put(entry.getKey(), entry.getValue().unwrapped().toString());
```
Comment (autumnust) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/retention/DatasetCleaner.java:105]: Does it work in this way ? It seems it will only get the top-level key and here it just convert the ConfigValue into a string which could itself be a hierarchical config.

File: gobblin-modules/gobblin-kafka-09/src/test/java/org/apache/gobblin/kafka/writer/Kafka09DataWriterTest.java (hunk 1 above)
Comment (ZihanLi58) [reply gobblin-modules/gobblin-kafka-09/src/test/java/org/apache/gobblin/kafka/writer/Kafka09DataWriterTest.java:87]: First time this test hang on travis build, so I made this change to see if it helps pass the travis build

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/retention/DatasetCleaner.java (hunk 2 above)
Comment (ZihanLi58) [reply gobblin-data-management/src/main/java/org/apache/gobblin/data/management/retention/DatasetCleaner.java]: It was working well magically... But I change the code to avoid confusion

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/retention/dataset/CleanableDatasetBase.java [hunk 3]
```diff
@@ -122,6 +122,6 @@ public CleanableDatasetBase(FileSystem fs, Properties properties, boolean simula
   @Override
   public List<VersionFinderAndPolicy<T>> getVersionFindersAndPolicies() {
     return ImmutableList
-        .<VersionFinderAndPolicy<T>> of(new VersionFinderAndPolicy<>(getRetentionPolicy(), getVersionFinder()));
+        .<VersionFinderAndPolicy<T>> of(new VersionFinderAndPolicy<>(getRetentionPolicy(), getVersionFinder(), null));
```
Comment (autumnust) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/retention/dataset/CleanableDatasetBase.java:125]: Shall we change it to ConfigFactory.empty() ?
